sound/soc/sof/topology.c

```diff
@@ -3515,6 +3515,45 @@ static int sof_set_pipe_widget(struct snd_sof_dev *sdev, struct snd_sof_widget *
 	return 0;
 }
 
+/*
+ * Function to get the initial value of SOF kcontrols. The value will be stored in
+ * the private data in struct snd_sof_control. This is particularly useful for binary kcontrols
+ * associated with static pipeline widgets to get the data size from the DSP.
+ */
+static int snd_sof_cache_kcontrol_val(struct snd_soc_component *scomp)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_control *scontrol;
+	int ipc_cmd, ctrl_type;
+	int ret = 0;
+
+	list_for_each_entry(scontrol, &sdev->kcontrol_list, list) {
+		struct snd_sof_widget *swidget;
+
+		swidget = snd_sof_find_swidget_by_id(scomp, scontrol->comp_id);
+		if (!swidget) {
+			dev_warn(scomp->dev, "No widget found with ID %d for control\n",
+				 scontrol->comp_id);
```

Comment:
nit-pick: move this to the helper?

Reply:
then it will have to be an error. is that ok?

Reply:
If you keep the print here then you can also print the `scontrol->name` perhaps to give more information?

Reply:
From the look, this should never happen as if we have something in the kcontrol_list then we must have the corresponding swidget in the widget_list, right?
But if we don't find a widget for a control then it is an error, no? Can it happen?